src/message.rs

```diff
@@ -0,0 +1,129 @@
+use crate::Hasher;
+
+pub use fuel_types::Bytes32;
+
+use core::fmt;
+use core::ops::Deref;
+
+/// Normalized signature message
+#[derive(Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(
+    feature = "serde-types-minimal",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+#[repr(transparent)]
+pub struct Message(Bytes32);
+
+impl Message {
+    /// Memory length of the type
+    pub const LEN: usize = Bytes32::LEN;
+
+    /// Normalize a message for signature
+    pub fn new<M>(message: M) -> Self
+    where
+        M: AsRef<[u8]>,
+    {
+        Self(Hasher::hash(message))
+    }
+
+    /// Add a conversion from arbitrary slices into owned
+    ///
+    /// # Safety
+    ///
+    /// There is no guarantee the provided bytes will be the product of a cryptographically secure
+    /// hash. Using insecure messages might compromise the security of the signature.
+    pub unsafe fn from_bytes_unchecked(bytes: [u8; Self::LEN]) -> Self {
+        Self(bytes.into())
+    }
+
+    /// Add a conversion from arbitrary slices into owned
+    ///
+    /// # Safety
+    ///
+    /// This function will not panic if the length of the slice is smaller than
+    /// `Self::LEN`. Instead, it will cause undefined behavior and read random
+    /// disowned bytes.
+    ///
+    /// This function extends the unsafety of [`Self::from_bytes_unchecked`].
+    pub unsafe fn from_slice_unchecked(bytes: &[u8]) -> Self {
+        Self(Bytes32::from_slice_unchecked(bytes))
+    }
+
+    /// Copy-free reference cast
+    ///
+    /// # Safety
+    ///
+    /// Inputs smaller than `Self::LEN` will cause undefined behavior.
+    ///
+    /// This function extends the unsafety of [`Self::from_bytes_unchecked`].
+    pub unsafe fn as_ref_unchecked(bytes: &[u8]) -> &Self {
+        // The interpreter will frequently make references to keys and values using
+        // logically checked slices.
+        //
+        // This function will avoid unnecessary copy to owned slices for the interpreter
+        // access
+        &*(bytes.as_ptr() as *const Self)
+    }
+}
+
+impl Deref for Message {
```

Comment:
All of these new-type passthrough methods (Deref, AsRef, Into, From, LowerHex, UpperHex, etc) could be automatically generated using the [derive_more](https://jeltef.github.io/derive_more/derive_more/) crate.

> it supports no_std out of the box

Reply:
https://github.com/FuelLabs/fuel-crypto/issues/6